packages/monaco/src/browser/monaco-workspace.ts

```diff
@@ -264,14 +264,27 @@ export class MonacoWorkspace implements lang.Workspace {
     async applyBulkEdit(workspaceEdit: monaco.languages.WorkspaceEdit, options?: EditorOpenerOptions): monaco.Promise<monaco.editor.IBulkEditResult> {
         let totalEdits = 0;
         let totalFiles = 0;
-        const uri2Edits = this.groupEdits(workspaceEdit);
-        for (const uri of uri2Edits.keys()) {
-            const editorWidget = await this.editorManager.open(new URI(uri), options);
+        let versionConflict = false;
+        const doc2Edits = this.groupEdits(workspaceEdit);
+        const editor2Edits = new Map<{ editor: MonacoEditor, model: monaco.editor.ITextModel }, monaco.languages.TextEdit[]>();
+        for (const doc of doc2Edits.keys()) {
+            const editorWidget = await this.editorManager.open(new URI(doc.uri), options);
             const editor = MonacoEditor.get(editorWidget);
             if (editor) {
                 const model = editor.document.textEditorModel;
+                if (doc.version === undefined || model.getVersionId() === doc.version) {
+                    editor2Edits.set({ editor, model }, doc2Edits.get(doc)!);
+                } else {
+                    versionConflict = true;
+                    break;
```

Comment:
This loop runs on a per-editor basis, so would it happen that only one of the editors was dirty?

In which case, is it better to break from the whole process, or rather `continue` which would skip this specific editor?

Reply:
I am strongly in favor of rejecting the entire workspaceEdit, as it is usually an atomic thing that ensures consistency. It's a pity we don't have a better channel for reporting errors, but cross-file operations are only poorly supported in Monaco/LSP.